pkg/controller/direct/bigquerydatatransfer/bigquerydatatransferconfig_controller.go

```diff
@@ -0,0 +1,351 @@
+	}
+	if m.config.UserProjectOverride && m.config.BillingProject != "" {
+		opts = append(opts, option.WithQuotaProject(m.config.BillingProject))
+	}
+
+	gcpClient, err := gcp.NewRESTClient(ctx, opts...)
+	if err != nil {
+		return nil, fmt.Errorf("building bigquerydatatransfer client: %w", err)
+	}
+	return gcpClient, err
+}
+
+func (m *model) AdapterForObject(ctx context.Context, reader client.Reader, u *unstructured.Unstructured) (directbase.Adapter, error) {
+	obj := &krm.BigQueryDataTransferConfig{}
+	if err := runtime.DefaultUnstructuredConverter.FromUnstructured(u.Object, &obj); err != nil {
+		return nil, fmt.Errorf("error converting to %T: %w", obj, err)
+	}
+
+	// Get ResourceID
+	resourceID := direct.ValueOf(obj.Spec.ResourceID)
+	if resourceID == "" {
+		serviceGeneratedID, err := parseServiceGeneratedIDFromExternalRef(obj)
+		if err != nil {
+			return nil, err
+		}
+		resourceID = serviceGeneratedID
+	}
+	// TODO(jingyih): need to figure out how to deal with resourceID. It is service generated.
+
+	location := obj.Spec.Location
+	if location == "" {
+		return nil, fmt.Errorf("cannot resolve location")
+	}
+
+	// Resolve PubSubTopic Ref
+	if obj.Spec.PubSubTopicRef != nil {
+		topic, err := refv1beta1.ResolvePubSubTopic(ctx, reader, obj, obj.Spec.PubSubTopicRef)
+		if err != nil {
+			return nil, err
+		}
+		obj.Spec.PubSubTopicRef.External = topic.String()
+	}
+
+	// Resolve BigQueryDataSet Ref
+	if obj.Spec.DatasetRef != nil {
+		dataset, err := refv1beta1.ResolveBigQueryDataset(ctx, reader, obj, obj.Spec.DatasetRef)
+		if err != nil {
+			return nil, err
+		}
+		obj.Spec.DatasetRef.External = dataset.String()
+	}
+
+	// Resolve KMSCryptoKey Ref
+	if obj.Spec.EncryptionConfiguration != nil {
+		kmsCryptoKeyRef, err := refv1beta1.ResolveKMSCryptoKeyRef(ctx, reader, obj, obj.Spec.EncryptionConfiguration.KmsKeyRef)
+		if err != nil {
+			return nil, err
+		}
+		obj.Spec.EncryptionConfiguration.KmsKeyRef = kmsCryptoKeyRef
+	}
+
+	// Resolve Project Ref
+	projectRef, err := refv1beta1.ResolveProject(ctx, reader, obj, obj.Spec.ProjectRef)
+	if err != nil {
+		return nil, err
+	}
+	projectID := projectRef.ProjectID
+	if projectID == "" {
+		return nil, fmt.Errorf("cannot resolve project")
+	}
+
+	var id *BigQueryDataTransferConfigIdentity
+	externalRef := direct.ValueOf(obj.Status.ExternalRef)
+	if externalRef == "" {
+		id = BuildID(projectID, location, resourceID)
+	} else {
+		id, err = asID(externalRef)
+		if err != nil {
+			return nil, err
+		}
+
+		if id.projectID != projectID {
+			return nil, fmt.Errorf("BigQueryDataTransferConfig %s/%s has spec.projectRef changed, expect %s, got %s",
+				u.GetNamespace(), u.GetName(), id.projectID, projectID)
+		}
+		if id.location != location {
+			return nil, fmt.Errorf("BigQueryDataTransferConfig %s/%s has spec.location changed, expect %s, got %s",
+				u.GetNamespace(), u.GetName(), id.location, location)
+		}
+		if id.transferConfigID != resourceID {
```

Comment:
I think this only works for acquire but not KCC-created.  Because the transferConfigID is server generated and KCC does not write it back to spec.resourceID (and not metadata.name)

Reply:
Mind elaborating a little bit more on why "we want users to specifically set spec.resourceID to the service generated ID"?

Reply:
I was referring to 
https://screenshot.googleplex.com/6Dcx5KjqfXvuexB

I think it makes sense to have users explicitly set `spec.ResourceID` to express the intend to "acquire a resource".